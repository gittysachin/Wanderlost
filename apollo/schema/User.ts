import { objectType, enumType, extendType, stringArg, idArg, arg } from '@nexus/schema'
import bcrypt from 'bcryptjs'
import jwt from 'jsonwebtoken'
import sgMail from '@sendgrid/mail'
import { promisify } from 'util'
import { randomBytes } from 'crypto'
import { setCookie } from '../cookies'

export const User = objectType({
  name: 'User',
  definition(t) {
    t.id('id')
    t.string('name')
    t.string('email')
    t.string('password')
    t.string('resetToken', { nullable: true })
    t.float('resetTokenExpiry', { nullable: true })
    t.field('permissions', {
      type: 'Permission',
      list: [false],
      nullable: true,
    })
    t.list.field('cart', { type: 'CartItem' })
    t.field('orders', {
      type: 'OrderItem',
      list: [false],
      nullable: true,
    })
  },
})

export const UserInput = objectType({
  name: 'UserInput',
  definition(t) {
    t.string('email')
    t.string('password')
    t.string('name')
  },
})

export const SuccessMessage = objectType({
  name: 'SuccessMessage',
  definition(t) {
    t.string('message')
  },
})

export const Permission = enumType({
  name: 'Permission',
  members: ['ADMIN', 'ITEMCREATE', 'ITEMDELETE', 'ITEMUPDATE', 'PERMISSIONUPDATE', 'USER'],
})

export const UserQuery = extendType({
  type: 'Query',
  definition(t) {
    t.field('user', {
      type: User,
      nullable: true,
      resolve(_root, _args, ctx) {
        if (!ctx.req.userId) {
          return null
        }
        return ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
          include: { cart: true },
        })
      },
    })

    t.field('users', {
      type: User,
      list: [false],
      nullable: true,
      resolve(_root, _args, ctx) {
        return ctx.prisma.user.findMany({
          include: { cart: true },
        })
      },
    })
  },
})

export const UserMutation = extendType({
  type: 'Mutation',
  definition(t) {
    t.field('signup', {
      type: User,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
        name: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        const user = await ctx.prisma.user.create({
          data: {
            email: args.email.toLocaleLowerCase(),
            name,
            password: await bcrypt.hash(args.password, 10),
            permissions: { set: ['USER', 'ITEMCREATE', 'ITEMDELETE', 'ITEMUPDATE'] },
          },
        })
        const token = jwt.sign({ userId: user.id }, process.env.APP_SECRET as string)
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365, // 1 year cookie
        })
        return user
      },
    })

    t.field('signin', {
      type: User,
      nullable: true,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if there is a user with that email
        const user = await ctx.prisma.user.findOne({ where: { email: args.email } })
        if (!user) {
          throw new Error(`No such user found for email ${args.email}`)
        }
        // 2. Check if their password is correct
        const valid = await bcrypt.compare(args.password, user.password)
        if (!valid) {
          throw new Error('Invalid Password!')
        }

        // 3. generate the JWT Token
        const token = jwt.sign({ userId: user.id }, process.env.APP_SECRET as string)

        // 4. Set the cookie with the token
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })
        // res.end(res.getHeader('Set-Cookie'))

        // 5. Return the user
        return user
      },
    })

    t.field('signout', {
      type: SuccessMessage,
      resolve(_root, _args, ctx) {
        setCookie(ctx.res, 'token', '', {})
        return { message: 'Goodbye!' }
      },
    })

    t.field('requestReset', {
      type: SuccessMessage,
      args: {
        email: stringArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        const user = await ctx.prisma.user.findOne({ where: { email: args.email } })
        if (!user) {
          throw new Error(`No user found for email: ${args.email}'`)
        }
        const randomBytesPromiseified = promisify(randomBytes)
        const resetToken = (await randomBytesPromiseified(20)).toString('hex')
        const resetTokenExpiry = Date.now() + 3600000 // 1 hour from now
        await ctx.prisma.user.update({
          where: { email: args.email },
          data: { resetToken, resetTokenExpiry },
        })
        const makeANiceEmail = (text: string) => `
          <div classname="email" style="
            border: 1px solid black;
            padding: 20px;
            font-family: sans-serif;
            line-height: 2;
            font-size: 20px;
          ">
            <h2>Hello there</>
            <p>${text}</p>
      
            <p>ðŸ˜˜, Davy Hausser</p>
          </div?
        `

        // TODO: make a nice email with functional link
        const msg = {
          to: user.email,
          from: 'davy.hausser@icloud.com',
          subject: 'Your Password Reset Token',
          text: 'and easy to do anywhere, even with Node.js',
          html: `<strong>${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}</strong>`,
          // html: makeANiceEmail(`Your Password Reset Token is here!
          // \n\n
          // <a href="${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}">Click Here to Reset</a>`),
        }
        sgMail.setApiKey(process.env.SENDGRID_API_KEY)
        sgMail.send(msg)
        return { message: 'Thanks!' }
      },
    })

    t.field('resetPassword', {
      type: User,
      args: {
        resetToken: stringArg({ required: true }),
        password: stringArg({ required: true }),
        confirmPassword: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if the passwords match
        if (args.password !== args.confirmPassword) {
          throw new Error("Yo Passwords don't match!")
        }
        // 2. check if its a legit reset token
        // 3. Check if its expired
        const [user] = await ctx.prisma.user.findMany({
          where: { resetToken: args.resetToken },
        })
        if (!user) {
          throw new Error('This token is either invalid or expired!')
        }
        // 4. Hash their new password
        const hash = await bcrypt.hash(args.password, 10)
        // 5. Save the new password to the user and remove old resetToken fields
        const updatedUser = await ctx.prisma.user.update({
          where: { email: user.email },
          data: {
            password: hash,
            resetToken: null,
            resetTokenExpiry: null,
          },
        })
        // 6. Generate JWT
        const token = jwt.sign({ userId: updatedUser.id }, process.env.APP_SECRET as string)
        // 7. Set the JWT cookie
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })
        // 8. return the new user
        return updatedUser
      },
    })

    t.field('updatePermissions', {
      type: User,
      nullable: true,
      args: {
        permissions: arg({
          type: 'Permission',
          list: [false],
        }),
        userId: idArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        // 1. Check if they are logged in
        if (!ctx.req.userId) {
          throw new Error('You must be logged in!')
        }
        // 3. Check if they have permissions to do this
        const user = await ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
        })
        hasPermission(user, ['ADMIN', 'PERMISSIONUPDATE'])
        // 4. Update the permissions
        return ctx.prisma.user.update({
          data: {
            permissions: { set: args?.permissions },
          },
          where: {
            id: ctx.req.userId,
          },
        })
      },
    })
  },
})

const hasPermission = (user: any, permissionsNeeded: any) => {
  const matchedPermissions = user.permissions.filter((permissionTheyHave: any) =>
    permissionsNeeded.includes(permissionTheyHave)
  )
  if (!matchedPermissions.length) {
    throw new Error(`You do not have sufficient permissions
      : ${permissionsNeeded}
      You Have:
      ${user.permissions}
      `)
  }
}
