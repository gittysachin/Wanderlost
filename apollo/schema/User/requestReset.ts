import sgMail from '@sendgrid/mail'
import { promisify } from 'util'
import { randomBytes } from 'crypto'

export async function requestReset(_, { email }, { prisma }) {
  const user = await prisma.user.findOne({ where: { email } })
  if (!user) {
    throw new Error(`No user found for email: ${email}'`)
  }
  const randomBytesPromiseified = promisify(randomBytes)
  const resetToken = (await randomBytesPromiseified(20)).toString('hex')
  const resetTokenExpiry = Date.now() + 3600000 // 1 hour from now
  await prisma.user.update({
    where: { email },
    data: { resetToken, resetTokenExpiry },
  })

  // TODO: make a nice email with functional link
  const msg = {
    to: user.email,
    from: 'davy.hausser@icloud.com',
    subject: 'Your Password Reset Token',
    text: 'and easy to do anywhere, even with Node.js',
    html: makeANiceEmail(`Your Password Reset Token is here!
    \n\n
    <a href="${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}">Click Here to Reset</a>`),
  }
  sgMail.setApiKey(process.env.SENDGRID_API_KEY)
  sgMail.send(msg)
  return { message: 'Thanks!' }
}

const makeANiceEmail = (text: string) => `
<div classname="email" style="
  border: 1px solid black;
  padding: 20px;
  font-family: sans-serif;
  line-height: 2;
  font-size: 20px;
">
  <h2>Hello there</>
  <p>${text}</p>

  <p>ðŸ˜˜, Davy Hausser</p>
</div?
`
