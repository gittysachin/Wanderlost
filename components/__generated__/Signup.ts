/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: Signup
// ====================================================

export interface Signup_signup {
  __typename: 'User'
  id: string
  email: string
  name: string
}

export interface Signup {
  signup: Signup_signup
}

export interface SignupVariables {
  email: string
  name: string
  password: string
}
