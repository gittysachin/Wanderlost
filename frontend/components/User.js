import { useQuery, gql } from '@apollo/client';

const CURRENT_USER_QUERY = gql`
  query {
    authenticatedUser {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  if (data) {
    return data.authenticatedUser;
  }
  return null;
}

export { CURRENT_USER_QUERY, useUser };
