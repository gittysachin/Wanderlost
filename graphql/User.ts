import {
  objectType,
  enumType,
  extendType,
  stringArg,
  idArg,
  arg,
} from '@nexus/schema'
import bcrypt from 'bcryptjs'
import jwt from 'jsonwebtoken'
import sgMail from '@sendgrid/mail'
import { promisify } from 'util'
import { randomBytes } from 'crypto'
import { setCookie, hasPermission } from './utils'

// import { Permission as PermissionType } from '@prisma/client'
// type Permissions = "ADMIN" | "ITEMCREATE" | "ITEMDELETE" | "ITEMUPDATE" | "PERMISSIONUPDATE" | "USER"

interface PermissionType {
  USER: 'USER'
  ADMIN: 'ADMIN'
  ITEMCREATE: 'ITEMCREATE'
  ITEMUPDATE: 'ITEMUPDATE'
  ITEMDELETE: 'ITEMDELETE'
  PERMISSIONUPDATE: 'PERMISSIONUPDATE'
}

export const User = objectType({
  name: 'User',
  definition(t) {
    t.id('id')
    t.string('name')
    t.string('email')
    t.string('password')
    t.string('resetToken', { nullable: true })
    t.float('resetTokenExpiry', { nullable: true })
    t.field('permissions', {
      type: 'Permission',
      list: [false],
      nullable: true,
    })
    t.list.field('cart', {
      type: 'CartItem',
      async resolve(root, _args, ctx) {
        const cart = await ctx.prisma.user
          .findOne({
            where: { id: root.id },
          })
          .cart()
        if (!cart) {
          throw new Error(`No cart found for user with id: ${root.id}`)
        }
        return cart
      },
    })
    t.field('orders', {
      type: 'OrderItem',
      list: [false],
      nullable: true,
      async resolve(root, _args, ctx) {
        const orders = await ctx.prisma.user
          .findOne({
            where: { id: root.id },
          })
          .OrderItem()
        if (!orders) {
          throw new Error(`No orders found for user with id: ${root.id}`)
        }
        return orders
      },
    })
  },
})

export const UserInput = objectType({
  name: 'UserInput',
  definition(t) {
    t.string('email')
    t.string('password')
    t.string('name')
  },
})

export const SuccessMessage = objectType({
  name: 'SuccessMessage',
  definition(t) {
    t.string('message')
  },
})

export const Permission = enumType({
  name: 'Permission',
  members: [
    'ADMIN',
    'ITEMCREATE',
    'ITEMDELETE',
    'ITEMUPDATE',
    'PERMISSIONUPDATE',
    'USER',
  ],
})

export const UserQuery = extendType({
  type: 'Query',
  definition(t) {
    t.field('user', {
      type: User,
      nullable: true,
      resolve(_root, _args, ctx) {
        if (!ctx.req.userId) {
          return null
        }
        return ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
          include: { cart: true },
        })
      },
    })

    t.field('users', {
      type: User,
      list: [false],
      nullable: true,
      resolve(_root, _args, ctx) {
        return ctx.prisma.user.findMany({
          include: { cart: true },
        })
      },
    })
  },
})

export const UserMutation = extendType({
  type: 'Mutation',
  definition(t) {
    t.field('signup', {
      type: User,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
        name: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        const user = await ctx.prisma.user.create({
          data: {
            email: args.email.toLocaleLowerCase(),
            name: args.name,
            password: await bcrypt.hash(args.password, 10),
            permissions: {
              set: ['USER', 'ITEMCREATE', 'ITEMDELETE', 'ITEMUPDATE'],
            },
          },
        })
        const token = jwt.sign(
          { userId: user.id },
          process.env.APP_SECRET as string
        )
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365, // 1 year cookie
        })
        return user
      },
    })

    t.field('signin', {
      type: User,
      nullable: true,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if there is a user with that email
        const user = await ctx.prisma.user.findOne({
          where: { email: args.email },
        })
        if (!user) {
          throw new Error(`No such user found for email ${args.email}`)
        }
        // 2. Check if their password is correct
        const valid = await bcrypt.compare(args.password, user.password)
        if (!valid) {
          throw new Error('Invalid Password!')
        }

        // 3. generate the JWT Token
        const token = jwt.sign(
          { userId: user.id },
          process.env.APP_SECRET as string
        )

        // 4. Set the cookie with the token
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })

        // 5. Return the user
        return user
      },
    })

    t.field('signout', {
      type: SuccessMessage,
      resolve(_root, _args, ctx) {
        setCookie(ctx.res, 'token', '', {})
        return { message: 'Goodbye!' }
      },
    })

    t.field('requestReset', {
      type: SuccessMessage,
      args: {
        email: stringArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        const user = await ctx.prisma.user.findOne({
          where: { email: args.email },
        })
        if (!user) {
          throw new Error(`No user found for email: ${args.email}'`)
        }
        const randomBytesPromiseified = promisify(randomBytes)
        const resetToken = (await randomBytesPromiseified(20)).toString('hex')
        const resetTokenExpiry = Date.now() + 3600000
        await ctx.prisma.user.update({
          where: { email: args.email },
          data: { resetToken, resetTokenExpiry },
        })

        const makeANiceEmail = (text: string) => `
          <div classname="email" style="
            border: 1px solid black;
            padding: 20px;
            font-family: sans-serif;
            line-height: 2;
            font-size: 20px;
          ">
            <h2>Hello there</>
            <p>${text}</p>

            <p>ðŸ˜˜, Davy Hausser</p>
          </div?
        `

        const msg = {
          to: user.email,
          from: 'davy.hausser@icloud.com',
          subject: 'Your Password Reset Token',
          text: 'and easy to do anywhere, even with Node.js',
          html: makeANiceEmail(`Your Password Reset Token is here!
          \n\n
          <a href="${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}">Click Here to Reset</a>`),
        }
        sgMail.setApiKey(process.env.SENDGRID_API_KEY as string)
        sgMail.send(msg)
        return { message: 'Thanks!' }
      },
    })

    t.field('resetPassword', {
      type: User,
      args: {
        resetToken: stringArg({ required: true }),
        password: stringArg({ required: true }),
        confirmPassword: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if the passwords match
        if (args.password !== args.confirmPassword) {
          throw new Error("Yo Passwords don't match!")
        }
        // 2. check if its a legit reset token
        // 3. Check if its expired
        const [user] = await ctx.prisma.user.findMany({
          where: { resetToken: args.resetToken },
        })
        if (!user) {
          throw new Error('This token is either invalid or expired!')
        }
        // 4. Hash their new password
        const hash = await bcrypt.hash(args.password, 10)
        // 5. Save the new password to the user and remove old resetToken fields
        const updatedUser = await ctx.prisma.user.update({
          where: { email: user.email },
          data: {
            password: hash,
            resetToken: null,
            resetTokenExpiry: null,
          },
        })
        // 6. Generate JWT
        const token = jwt.sign(
          { userId: updatedUser.id },
          process.env.APP_SECRET as string
        )
        // 7. Set the JWT cookie
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })
        // 8. return the new user
        return updatedUser
      },
    })

    t.field('updatePermissions', {
      type: User,
      nullable: true,
      args: {
        permissions: arg({
          type: 'Permission',
          list: [false],
        }),
        userId: idArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        // 1. Check if they are logged in
        if (!ctx.req.userId) {
          throw new Error('You must be logged in!')
        }
        // 2. Retrieve user from the database
        const user = await ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
        })
        // 3. Check if they have permissions to do this
        if (!user) throw new Error(`User not found: ${ctx.req.userId}`)
        hasPermission(user, ['ADMIN', 'PERMISSIONUPDATE'])
        // 4. Update the permissions
        return ctx.prisma.user.update({
          data: {
            permissions: { set: args?.permissions as PermissionType[] },
          },
          where: {
            id: ctx.req.userId,
          },
        })
      },
    })

    t.field('updateUser', {
      type: 'User',
      args: {
        id: idArg({ required: true }),
        name: stringArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        if (!ctx.req.userId) {
          throw new Error('You must be logged in!')
        }
        return ctx.prisma.user.update({
          where: { id: args.id },
          data: {
            name: args.name,
          },
        })
      },
    })
  },
})
