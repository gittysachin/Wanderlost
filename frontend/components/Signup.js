import { useState } from 'react';
import { useMutation } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';
import Form from './styles/Form';
import Error from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`;

function Signup() {
  const [name, setName] = useState('');
  const [password, setPassword] = useState('');
  const [email, setEmail] = useState('');

  const [signup, { error, loading }] = useMutation(SIGNUP_MUTATION, {
    variables: { name, password, email },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  return (
    <Form
      method="post"
      onSubmit={async (e) => {
        e.preventDefault();
        const res = await signup();
        console.log(res);
        setName('');
        setPassword('');
        setEmail('');
      }}
    >
      <fieldset disabled={loading} aria-busy={loading}>
        <h2>Sign Up for an Account</h2>
        <Error error={error} />
        <label htmlFor="email">
                Email
          <input
            type="email"
            name="email"
            placeholder="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
        </label>
        <label htmlFor="name">
                Name
          <input
            type="text"
            name="name"
            placeholder="name"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </label>
        <label htmlFor="password">
                Password
          <input
            type="password"
            name="password"
            placeholder="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </label>

        <button type="submit">Sign Up!</button>
      </fieldset>
    </Form>
  );
}

export default Signup;
