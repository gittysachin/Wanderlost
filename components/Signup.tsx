import { useMutation, gql } from '@apollo/client'
import { useForm } from 'react-hook-form'
import Form from './styles/Form'
import Error, { FormError } from './ErrorMessage'
import { CURRENT_USER_QUERY } from './User'
import { Signup as SignupTypes, SignupVariables } from './__generated__/Signup'

const SIGNUP_MUTATION = gql`
  mutation Signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`

function Signup(): JSX.Element {
  const defaultValues = { email: '', name: '', password: '' }
  const { register, handleSubmit, reset, errors } = useForm({ defaultValues })
  const [signup, { error, loading, data }] = useMutation<
    SignupTypes,
    SignupVariables
  >(SIGNUP_MUTATION, { refetchQueries: [{ query: CURRENT_USER_QUERY }] })

  async function onSubmit(data: SignupVariables) {
    await signup({
      variables: {
        email: data.email,
        name: data.name,
        password: data.password,
      },
    })
    reset(defaultValues)
  }

  return (
    <Form method="post" onSubmit={handleSubmit(onSubmit)}>
      <fieldset disabled={loading} aria-busy={loading} data-testid="loading">
        {data && data.signup && (
          <p>Signed up with {data.signup.email} â€” Please Sign In now</p>
        )}
        <h2>Sign Up for an Account</h2>
        <Error error={error} />
        <label>
          Email
          <input
            type="email"
            name="email"
            placeholder="email"
            autoComplete="email"
            ref={register({ required: true })}
          />
        </label>
        <FormError error={errors.email} />
        {errors.email && <p>This is required</p>}

        <label>
          Name
          <input
            type="text"
            name="name"
            placeholder="name"
            autoComplete="name"
            ref={register({ required: true })}
          />
        </label>
        <FormError error={errors.name} />

        <label>
          Password
          <input
            type="password"
            name="password"
            placeholder="password"
            autoComplete="new-password"
            ref={register({ required: true })}
          />
        </label>
        <FormError error={errors.password} />

        <button type="submit">Sign Up!</button>
      </fieldset>
    </Form>
  )
}

export default Signup
export { SIGNUP_MUTATION, Signup }
