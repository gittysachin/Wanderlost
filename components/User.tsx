import { useQuery, gql } from '@apollo/client'
import { GetCurrentUser } from './__generated__/GetCurrentUser'

const CURRENT_USER_QUERY = gql`
  query GetCurrentUser {
    user {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`

function useUser() {
  const { data } = useQuery<GetCurrentUser>(CURRENT_USER_QUERY)
  if (data) {
    return data.user
  }
  return null
}

export { CURRENT_USER_QUERY, useUser }
