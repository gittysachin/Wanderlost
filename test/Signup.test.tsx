import {
  render,
  fireEvent,
  screen,
  waitFor,
  userEvent,
  fakeUser,
} from '../lib/test-utils'
import { SIGNUP_MUTATION, Signup } from '../components/Signup'
import { CURRENT_USER_QUERY } from '../components/User'

const me = fakeUser()
const mocks = [
  // signup mock mutation
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password: 'wes',
      },
    },
    result: {
      data: {
        signup: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // current user query mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedUser: me } },
  },
]

describe('<Signup/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(<Signup />)
    expect(container).toMatchSnapshot()
  })

  it('renders without error', () => {
    render(<Signup />, { mocks: [] })
  })

  it('should render loading state initially', async () => {
    const component = render(<Signup />, { mocks })

    // find the button and simulate a click
    const button = await component.findByRole('button')
    waitFor(() => {
      fireEvent.click(button) // fires the mutation
    })

    expect(screen.getByTestId('loading')).toHaveAttribute('aria-busy', 'true')
    expect(screen.getByTestId('loading')).not.toHaveAttribute('disabled')
  })

  it('calls the mutation properly', () => {
    render(<Signup />, { mocks })
    userEvent.type(screen.getByPlaceholderText('name'), me.name)
    userEvent.type(screen.getByPlaceholderText('email'), me.email)
    userEvent.type(screen.getByPlaceholderText('password'), 'wes')

    expect(screen.getByTestId('loading')).toHaveAttribute('aria-busy', 'false')
    expect(screen.getByTestId('loading')).not.toHaveAttribute('disabled')

    waitFor(() => {
      // userEvent.click(screen.getByText('Sign Up!'))
      fireEvent.click(screen.getByText('Sign Up!'))
    })

    // await new Promise((resolve) => setTimeout(resolve, 0)) // wait for response

    // waitFor(() => {
    // expect(screen.getByTestId('loading')).toHaveAttribute('aria-busy', 'true')
    // expect(screen.getByTestId('loading')).toHaveAttribute('disabled')
    // })

    // loading state
    // waitFor(() => {
    //   expect(screen.getByTestId('loading')).toHaveAttribute('aria-busy', 'true')
    //   expect(screen.getByTestId('loading')).toHaveAttribute('disabled')
    // })
    // screen.findByText(`Signed up with ${me.email} â€” Please Sign In now`)
  })
})
