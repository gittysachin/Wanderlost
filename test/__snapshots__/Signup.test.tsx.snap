// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-bdnylx eviPWj"
    method="post"
  >
    <fieldset
      aria-busy="false"
    >
      <h2>
        Sign Up for an Account
      </h2>
      <label>
        Email
      </label>
      <input
        autocomplete="email"
        name="email"
        placeholder="email"
        type="email"
      />
      <label>
        Name
      </label>
      <input
        autocomplete="name"
        name="name"
        placeholder="name"
        type="text"
      />
      <label>
        Password
      </label>
      <input
        autocomplete="new-password"
        name="password"
        placeholder="password"
        type="password"
      />
      <button
        type="submit"
      >
        Sign Up!
      </button>
    </fieldset>
  </form>
</div>
`;
